include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td

```diff
@@ -16,7 +16,8 @@ class TTG_Op<string mnemonic, list<Trait> traits = []> :
     Op<TritonGPU_Dialect, mnemonic, traits>;
 
 def TTG_ConvertLayoutOp : TTG_Op<"convert_layout",
-                                 [SameOperandsAndResultShape, NoSideEffect]> {
+                                 [SameOperandsAndResultShape, 
+                                  MemoryEffects<[MemRead]>]> { // Potentially read shared memory 
```

Comment:
This part is probably tricky but safe? In what cases will we lose optimization opportunities?

Reply:
I can think of two problematic things with this:
(1) Practically, I've seen [MemRead] preventing MLIR's CSE from working properly (https://github.com/openai/triton/blob/triton-mlir/include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td#L82)
(2) Conceptually, the TritonGPU doc should not mention shared memory at all; this is an abstraction leak

Reply:
Should every op that doesn't touch global memory be defined with `NoSideEffect`?

Reply:
Ideally, yes. Except for AllocTensorOp

Reply:
Done. Took out MemRead and MemWrite on shared memory and still passed the tests.

> (1) Practically, I've seen [MemRead] preventing MLIR's CSE from working properly (https://github.com/openai/triton/blob/triton-mlir/include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td#L82)

Now it seems to work.

> (2) Conceptually, the TritonGPU doc should not mention shared memory at all; this is an abstraction leak

It's doable. We could replace `triton_gpu.alloc_tensor` with `bufferization.alloc_tensor` after upgrading LLVM. The current solution can be viewed as a temporary workaround.